pdns/misc.cc

```diff
@@ -225,6 +229,45 @@ auto pdns::getMessageFromErrno(const int errnum) -> std::string
   return message;
 }
 
+#if defined(HAVE_LIBCRYPTO)
+auto pdns::OpenSSL::error(const std::string& errorMessage) -> std::runtime_error
+{
+  unsigned long errorCode = 0;
+  auto fullErrorMessage{errorMessage};
+#if OPENSSL_VERSION_MAJOR >= 3
+  const char* filename = nullptr;
+  const char* functionName = nullptr;
+  int lineNumber = 0;
+  while ((errorCode = ERR_get_error_all(&filename, &lineNumber, &functionName, nullptr, nullptr)) != 0) {
+    const auto* lib = ERR_lib_error_string(errorCode);
+    const auto* reason = ERR_reason_error_string(errorCode);
+    fullErrorMessage += std::string(": ") + std::to_string(errorCode) + ":" + lib + "::" + reason;
+    if (filename != nullptr) {
+      fullErrorMessage += std::string(" - ") + filename;
+    }
+    if (lineNumber != 0) {
+      fullErrorMessage += std::string(":") + std::to_string(lineNumber);
+    }
+    if (functionName != nullptr) {
+      fullErrorMessage += std::string(" - ") + functionName;
+    }
+  }
+#else
+  while ((errorCode = ERR_get_error()) != 0) {
+    const auto* lib = ERR_lib_error_string(errorCode);
+    const auto* reason = ERR_reason_error_string(errorCode);
+    fullErrorMessage += std::string(": ") + std::to_string(errorCode) + ":" + lib + "::" + reason;
```

Comment:
Two questions:
- do we need to be careful about `lib` or `reason` being `nullptr`? The documentation in 1.1.1l states that it might happen: `ERR_lib_error_string(), ERR_func_error_string() and ERR_reason_error_string() return the strings, and B<NULL> if none is registered for the error code.`
- do we want `ERR_func_error_string()` as well, since it was still OK to use before 3.0.0 and might provide valuable information?

Reply:
The answer is yes to both. I've added the necessary code.

Reply:
Awesome, thanks!